src/transformStateWithClones.js

```diff
@@ -6,8 +6,31 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+'use strict';
+
+function transformStateWithClones(initialState, operations) {
+  const result = [];
+  let currentState = { ...initialState };
+
+  for (const operation of operations) {
+    if (operation.type === 'addProperties') {
+      currentState = { ...currentState, ...operation.extraData };
+    } else if (operation.type === 'removeProperties') {
+      currentState = Object.keys(currentState)
+        .filter((key) => !operation.keysToRemove.includes(key))
+        .reduce((newState, key) => {
+          newState[key] = currentState[key];
+
+          return newState;
+        }, {});
+    } else if (operation.type === 'clear') {
+      currentState = {};
+    }
+
+    result.push({ ...currentState });
```

Comment:
The `result.push({ ...currentState });` should be placed outside of the `if-else` block, but still within the loop, to ensure that the current state is pushed to the result array at the end of each loop cycle, after processing the current operation. This aligns with the additional prompt rule: 'It is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after current action is processed.'